group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntime.java

```diff
@@ -0,0 +1,1009 @@
+         * Reverts the last written offset. This also reverts the snapshot
+         * registry to this offset. All the changes applied after the offset
+         * are lost.
+         *
+         * @param offset The offset to revert to.
+         */
+        private void revertLastWrittenOffset(
+            long offset
+        ) {
+            if (offset > lastWrittenOffset) {
+                throw new IllegalStateException("New offset " + offset + " of " + tp +
+                    " must be smaller than " + lastWrittenOffset + ".");
+            }
+
+            log.debug("Revert last written offset of {} to {}.", tp, offset);
+            lastWrittenOffset = offset;
+            snapshotRegistry.revertToSnapshot(offset);
+        }
+
+        /**
+         * Updates the last committed offset. This completes all the deferred
+         * events waiting on this offset. This also cleanups all the snapshots
+         * prior to this offset.
+         *
+         * @param offset The new last committed offset.
+         */
+        private void updateLastCommittedOffset(
+            long offset
+        ) {
+            if (offset <= lastCommittedOffset) {
+                throw new IllegalStateException("New committed offset " + offset + " of " + tp +
+                    " must be larger than " + lastCommittedOffset + ".");
+            }
+
+            log.debug("Update committed offset of {} to {}.", tp, offset);
+            lastCommittedOffset = offset;
+            deferredEventQueue.completeUpTo(offset);
+            snapshotRegistry.deleteSnapshotsUpTo(offset);
+        }
+
+        /**
+         * Transitions to the new state.
+         *
+         * @param newState The new state.
+         */
+        private void transitionTo(
+            CoordinatorState newState
+        ) {
+            if (!newState.canTransitionFrom(state)) {
+                throw new IllegalStateException("Cannot transition from " + state + " to " + newState);
+            }
+
+            log.debug("Transition from {} to {}.", state, newState);
+            switch (newState) {
+                case LOADING:
+                    state = CoordinatorState.LOADING;
+                    coordinator = coordinatorBuilderSupplier
+                        .get()
+                        .withSnapshotRegistry(snapshotRegistry)
+                        .build();
+                    break;
+
+                case ACTIVE:
+                    state = CoordinatorState.ACTIVE;
+                    partitionWriter.registerListener(tp, highWatermarklistener);
+                    snapshotRegistry.getOrCreateSnapshot(0);
+                    coordinator.onLoaded();
+                    break;
+
+                case FAILED:
+                    state = CoordinatorState.FAILED;
+                    break;
+
+                case CLOSED:
+                    state = CoordinatorState.CLOSED;
+                    partitionWriter.deregisterListener(tp, highWatermarklistener);
```

Comment:
are all of these methods safe to call from initial state? seems like that state transition is valid.

Reply:
almost. updated the code.